### PR TITLE
plugins/gcp_cas: Update to v1 of the privatecas api

### LINKED ISSUE #2567: gcp_cas plugin crashing since GA of CAS
Recently the test branch of serverless identity issuance began crashing, we suspect that this is related to CAS going GA. We tested successfully with pre-GA CAS.

```INFO[0000] Data directory: "./.data_server"             
INFO[0000] Opening SQL database                          db_type=sqlite3 subsystem_name=sql
DEBU[0000] Code and DB schema versions are the same. No migration needed  schema=16 subsystem_name=sql version_info=1.0.0-dev-9b1d45c
INFO[0000] Connected to SQL database                     read_only=false subsystem_name=sql type=sqlite3 version=3.34.0
INFO[0000] Plugin loaded                                 external=false plugin_name=join_token plugin_type=NodeAttestor subsystem_name=catalog
INFO[0000] Plugin loaded                                 external=false plugin_name=disk plugin_type=KeyManager subsystem_name=catalog
WARN[0000] This plugin relies on GCP Certificate Authority Service which is currently in Beta  external=false plugin_name=gcp_cas plugin_type=UpstreamAuthority subsystem_name=catalog
INFO[0000] Plugin loaded                                 external=false plugin_name=gcp_cas plugin_type=UpstreamAuthority subsystem_name=catalog
DEBU[0000] Loading journal                               path=.data_server/journal.pem subsystem_name=ca_manager
INFO[0000] Journal loaded                                jwt_keys=3 subsystem_name=ca_manager x509_cas=0
DEBU[0000] Preparing X509 CA                             slot=A subsystem_name=ca_manager
DEBU[0000] Request to GCP_CAS to mint new X509           external=false plugin_name=gcp_cas plugin_type=UpstreamAuthority subsystem_name=catalog
ERRO[0000] Unable to rotate X509 CA                      error="rpc error: code = InvalidArgument desc = upstreamauthority(gcp_cas): no certificate authorities found with label pair \"test\":\"test\"" subsystem_name=ca_manager
INFO[0000] JWT key activated                             expiration="2021-09-24T17:10:08Z" issued_at="2021-09-23T17:10:08Z" slot=B subsystem_name=ca_manager
ERRO[0000] Fatal run error                               error="rpc error: code = InvalidArgument desc = upstreamauthority(gcp_cas): no certificate authorities found with label pair \"test\":\"test\""
ERRO[0000] Server crashed                                error="rpc error: code = InvalidArgument desc = upstreamauthority(gcp_cas): no certificate authorities found with label pair \"test\":\"test\""
```

The plugin seems to not be able to find the CA based on labels anymore, which previously worked. Note: we verified the labels are in fact attached to the CA. Also the contents are renamed here to "test" - the actual values are different.

* **Version**: 

To reproduce:

`git clone -b agentless_prop git@github.com:MarcosDY/spire.git`

SHA: 16bb3894490700e679de9c18390527aa41f060d5

* **Platform**: 

Ubuntu docker image

* **Subsystem**: 

spire-server gcp_cas plugin for serverless

### PR DESCRIPTION
**Pull request by mikhailswift**
**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
Affects the server/upstreamauthority/gcp_cas plugin.  Adds new config field which will require intervention from users of the plugin.

**Description of change**
Google's CAS service entered general availability and in doing so v1 of
the API was published.  When this happened requests to the v1beta1
seemingly stopped returning data.

* Adds `ca_pool` to the plugin config. CAs now live within CA Pools in
  CAS, so we need to know which pool to search in for CA.
* `ReusableConfigs` were removed from the `CreateCertificate` API call.
  The request now expects the `X509Config` argument to be provided,
  which details the Key Usage and other X509 options for the certificate
  to be created.

Signed-off-by: Mikhail Swift <mswift@mswift.dev>

**Which issue this PR fixes**
fixes #2567

### CODE REVIEW

**Review by amartinezfayo — COMMENTED** [review]
Thanks @mikhailswift for this fix!
I'm concerned about breaking backwards compatibility. Please see my comment below.

File: pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go
Comment (colek42) [reply pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go:145]: It looks like we could have the plugin create a CA pool, but I think that may be out of scope.  If the v1beta1 API is gone we really need to make the change.  Are other users on GCP having the same issue?

Comment (mikhailswift) [reply pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go:145]: I don't like the idea of having the plugin create the pool.  However there's a few different ways of varying efficiency I can think of to accomplish this:

1.  Search every ca pool in the project & region for a certificate authority with the matching labels
2.  Enforce that the ca pool must have the same labels and search for the ca pool with the defined labels and a root in the pool with the same labels.

Comment (mikhailswift) [reply pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go:145]: @amartinezfayo Any preference or other ideas from the maintainers on the preferred way to handle the backward compatibility issue?

Personally I think making ca_pool optional and otherwise searching through every CA pool in the provided region for a matching CA makes sense and acts in accordance with the beta functionality.

Comment (azdagron) [reply pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go:145]: I'm not sure how onerous that search is but that seems like a fine approach to me provided that we:
1. Fail if more than one CA is matched
2. Issue a warning when ca_pool is not configured

The warning could be something like:
```
// TODO: remove deprecated fallback behavior in 1.2
log.Warn("The ca_pool value is not configured. Falling back to searching the region for matching CAs. The ca_pool configurable will be required in a future release.")
```

Comment (azdagron) [reply pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go:145]: Actually, we can't fail if more than one is matched. That breaks the current behavior of adding the rest to the trust bundle.

Comment (azdagron) [reply pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go:145]: Or maybe we just fail if the set of matched CAs spans more than one pool?

Comment (azdagron) [reply pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go:145]: This seems to align pretty closely with what my understanding of a CA pool is. Also seems that at some later point we want to instead sign using a pool and not require labels, but I'm new to this GCP CAS thing.

Comment (amartinezfayo) [reply pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go:145]: Failing if the set of matched CAs spans more than one pool makes sense to me.

**Comment by azdagron** [comment]
@mikhailswift looks like we're missing DCO sign-off on the commit?

**Comment by mikhailswift** [comment]
@azdagron Sorry about that, didn't mean to push that commit yet.  There's some slight changes since my initial push and the DCO is now added.  Ready for review

**Review by azdagron — DISMISSED** [review]
Thanks @mikhailswift ! @amartinezfayo is also taking a look to ensure we don't miss anything.

**Review by azdagron — APPROVED** [review]

**Review by amartinezfayo — DISMISSED** [review]
LGTM! 🎉

**Review by azdagron — APPROVED** [review]

**Pull request merged**
